### ISSUE TITLE
What is the correct authorization to join a channel/bot via JS client? 

### ISSUE THREAD

**Issue by zontarian** [instedd/aida#276]
I posted this same issue on the `aida-ui` GitHub repo, maybe I was wrong, so please accept my apologies for cross posting [issue on aida-ui](https://github.com/instedd/aida-ui/issues/239)

Hi,
I'm trying to connect to a Bot I've created, using a JS client (so, websockets). This is my first go at it.
I'm using as a starting point
https://hexdocs.pm/phoenix/js/#phoenix
https://github.com/instedd/aida-ui/blob/master/app/javascript/components/ChatClient.jsx
 and instructions in 
https://github.com/instedd/aida/blob/master/docs/websocket.md

So far I've managed to open the Phoenix socket and connect to it, but when I try to join the channel Iike this
```    
let socket = new Socket('wss://aida.xxxxxx.org/socket');
socket.connect();
let channel = socket.channel('bot:'+botId, {'access_token': accessToken}); //<-- error here
```
I get

`{reason: "unauthorized"}`

`botId` is a number, the one I see in this URL `https://aida.xxxxxxxxx.org/b/10/channel` (10). While `accessToken` is the token I've written in the `Channel` tab of my Bot configuration, under the heading 
![screen shot 2018-09-19 at 15 46 29](https://user-images.githubusercontent.com/2872433/45757255-35f62080-bc23-11e8-9b1f-a3068e4a049e.png)

So, what is the correct authorization? 

----

After some tinkering woth Google DEv tools..  i got this

Well..it seems my parameters were wrong.
By debugging the chat in the "Behaviour" tab, I discovered
* a `uuid` for the bot. It's something like `ae9c0c65-xxx-xxxx-xxxx-xxxxxxxxxxx` (obfuscated..) so it seems it's not `10`
* an access token which is different again, in the same UUID hex form of 16bytes in hex. So it doesn't seem the access token inserted in the `channel` tab. 

With these two parameter my sample connection worked and I was able to join the channel.

Now the question is, how to infer these values from the info on the web site? :) Maybe there is a service to call beforehand.. 

Any help?

**Comment by nekron** [comment]
Hi @zontarian ! it's ok to create issues in any of the repos. We try to keep the issues related to the repo that will have the changes made, but in this case it's the same.

So to answer your questions:

Yes, the bot id is actually the bot uuid, as it's the id of the `aida` runtime (this repo) in contrast with the id of the front end setup client (`aida-ui`).

You can obtain it using the bots api (https://aida.xxxxxx.org/api/v1/bots.json?access_token=YOUR_API_TOKEN)

The access token will be the access token you set in the channels page, but using this bot id that I mentioned.

What you saw when inspecting the chat in the behavior tab is "another bot", a temporary one that is created for testing purposes. For `aida-ui`, this bot will be just a "preview-uuid" in the same bot table. But for `aida` it will be a totally different and independent bot that will be re-published every time you open that chat window.

The idea is to be able to test the bot using exactly the same runtime that will be used for production, but without overwriting the production bot until properly deployed.

So for every bot that you have in `aida-ui`, there will be two bots in the `aida` runtime.

The only difference between those bots is that the temporary one will be deployed with only one hardcoded websocket channel, with an autogenerated access token, that will be used to chat from the behavior tab. That's why you saw a second UUID used as access token.

What you did by connecting using that credentials, was to talk to the preview bot. And that's the reason why you probably didn't saw any change in the analytics tab after you talked to the bot either. Because those stats are targeting the primary bot.

If there is anything that we can help, please feel free to open another issue or reopen this one.

Best regards!

**Comment by zontarian** [comment]
Hi @nekron ,
Thanks for the detailed answer. 
That explains hat were the `preview-xxx` API calls and variables. It makes sense to have a testing bot.

I tried this URL: `https://aida.xxxxxx.org/api/v1/bots.json?access_token=token-in-channel-page`

but I get a JSON like this.. the bot Id is the numeric ID, in my case still `10`

```
[
  {
    "id": 10,
    "name": "xxxxxxxxx",
    "published": true,
    "channel_setup": false,
    "permissions": {
      "can_admin": false,
      "can_publish": true,
      "manages_behaviour": true,
      "manages_content": true,
      "manages_variables": true,
      "manages_results": true
    },
     "collaborator_id": 9
  }
]
```

So I'm still at the starting point, with a `10`. And using it to connect to che channel I still get `unauthorized`. 
 
I am missing something: to properly connect to the channel I don't need the numeric botId, rather the 16byte hex UUID, is it correct?
So, where do I get the bot UUID (16bytes in hex?). 
I am missing some point, obviously, could you please help me again? 

Thanks for your time.

**Comment by nekron** [comment]
Sorry, it seems that we changed that api in the latest release. I was checking against our staging server and I failed to realize that production was one version behind and was answering something completely different.

You can try it against our staging server at aida-stg.instedd.org, it should answer with something like this:

```
[
  {
    id: 9,
    name: "Bot 9",
    wit_ai_auth_token: "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
    published: true,
    channel_setup: true,
    uuid: "d0cf58e3-2e37-4aa3-a82d-xxxxxxxxxxxx",
    channels: [
      "Facebook",
      "Websocket"
    ],
    updated_at: "2018-09-18T14:59:33.068Z",
    active_users: 0,
    permissions: {
      can_admin: true,
      can_publish: true,
      manages_behaviour: true,
      manages_content: true,
      manages_variables: true,
      manages_results: true
    },
    collaborator_id: null
  },
  ...
]
```

The id that you are looking for is the `uuid: "d0cf58e3-2e37-4aa3-a82d-xxxxxxxxxxxx"`

**Comment by zontarian** [comment]
Hi Nekron,
this makes more sense.
Unfortunately hte UUID is null!

```
[
  {
    "id": 124,
    "name": "Bot 124",
    "wit_ai_auth_token": null,
    "published": false,
    "channel_setup": false,
    "uuid": null,
    "channels": [
        "Facebook",
        "Websocket"
    ],
    "updated_at": "2018-07-23T15:29:30.101Z",
    "active_users": 0,
    "permissions": {
      "can_admin": true,
      "can_publish": true,
      "manages_behaviour": true,
      "manages_content": true,
      "manages_variables": true,
      "manages_results": true
    },
    "collaborator_id": null
}
]
```

Besides this, could I use staging and be sure that in the short term the modifications will be brought over to production?
Should I create some Bots in staging?
Thanks

**Comment by zontarian** [comment]
Hi,
I've solved one problem: the bot 124 is broken so there is no UUID. Another Bot created shows a good UUID! Thanks!

But I have still two issues, one is blocking the other is not..
```
let socket = new Socket('wss://aida-stg.instedd.org/socket');
socket.connect();
let channel = socket.channel('bot:'+botUUID, {'access_token': accessToken}); //<-- error here
```
1. If I connect using the above syntax I still get `Unauthorized`. The access token is a textual one, as set in the WebSocket section of the Channel tab on the staging server. So maybe I am still doing something wrong.
2. the call to get BOT info, as suggested to you, `https://aida-stg.instedd.org/api/v1/bots.json?access_token=<MY_ACCESS_TOKEN>` needs authentication (status code 401). In a browser where I am authenticated in the staging server it works. Could you provide the specs of the authentication needed in order to call this API? 

Thanks so much for your help
W

**Comment by nekron** [comment]
Hi, sorry for the late response, let me answer your questions one by one.

- You can use staging for the time being. The changes should be in production in about a week.

- The reason you didn't have the uuid in that bot is because it was not published yet, so the runtime hadn't yet generated it (you also had the `published: false` there as a clue).

- Regarding the websocket connection problem, I'm not entirely sure what's wrong, but maybe this can help: we now have a stand alone client, so when you create a Web channel, it will provide you with an url to connect to it using our stand alone chat. The bot needs to be deployed first because right now the url will just contain a `null` in there. You can check its code in https://github.com/instedd/aida-ui/blob/master/app/javascript/components/WebChatWindow.jsx, https://github.com/instedd/aida-ui/blob/master/app/javascript/components/WebChatClient.jsx, and https://github.com/instedd/aida-ui/blob/master/app/javascript/components/WebChat.jsx

- Regarding the API, keep in mind that the token that you need to use is the one you get from the API page in the same server (in this case https://aida-stg.instedd.org/settings/api).
I've successfully tried both 
`curl -H "Authorization: Bearer <my_token>" https://aida-stg.instedd.org/api/v1/bots.json`
and 
`curl https://aida-stg.instedd.org/api/v1/bots.json\?access_token\=<my_token>`
And it all seems to be working.
Maybe you are using the wrong token?

Best!

**Comment by nekron** [comment]
The API generated tokens are bearer tokens provided by our oauth service (https://github.com/instedd/guisso). The staging server uses https://login-stg.instedd.org/access_tokens while the production server uses https://login.wfp.instedd.org/access_tokens. The tokens will never expire, but you can revoke them from there if you like.

The channel token should be the one specified there in the channel page (in staging it will be autogenerated from now on and provided there within the web chat url). I don't understand why can't you connect using it.

As a quick test I'd suggest you try to connect with our web chat using https://aida-stg.instedd.org/c/[bot_id]/[channel_auth_token] as provided in the web channel page. If that works, you should be able to connect using those credentials.

If the problem persists, feel free to drop me an email with the bot id and I'll try to provide you the right credentials.

**Comment by zontarian** [comment]
Hi Nekron,
Ok! This is very clear. I've seen the `tokens` page and since they are "unlimited" I can embed them in the code.

TLDR: *now it works*, maybe got needed to be published. Thanks!

I've been to the cannel/websocket page and seen the new URL that you provide. In this the credentials are the one I used. The bot seemed to be published since the url had no `null` inside.

But what I did was, by mistake, to press the "publish" button in the behaviour tab, and now it works.

To recap what I did is:

1. Use the URL `https://aida-stg.instedd.org/api/v1/bots.json?access_token=<MY ACCESS TOKEN FROM API>` to get the UUID of the Bot I want (let's call it `BOT_UUID`)
2. setup an access token in the `cannels/websocket` page, of my choice. let's call this one 'CHOSEN_ACCESS_TOKEN`
3. Import Phoenix channel library in a JS file
4. from JS, connect to channel with `bot`:`BOT_UUID` and `access_token`=`CHOSEN_ACCESS_TOKEN`

This works now! I can send and receive messages.! I don't know if this was due to the publishing action, or what..

Thanks a lot for your explanations and guidance,
Best!
